### PR TITLE
Tests for OnUnauthorizedClientConnect

### PR DESCRIPTION
**Pull request by sburmanoctopus**
# Background

There are currently no tests relating to the use of `OnUnauthorizedClientConnect` in Halibut.

# Results

Related to https://github.com/OctopusDeploy/Issues/issues/8266

## Before
No tests relating to using `OnUnauthorizedClientConnect`

## After

There are now tests for `OnUnauthorizedClientConnect`

### Duplicate Test Cases
We also found an issue with `LatestClientAndPreviousServiceVersionsTestCases` where it duplicated test cases. To help ensure this doesn't happen, we made the `ClientAndServiceTestCasesBuilder` use distinct list values.

# How to review this PR

Quality :heavy_check_mark:

# Pre-requisites

- [ ] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [ ] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [ ] I have considered appropriate testing for my change.

### CODE REVIEW

File: source/Halibut.Tests/BadCertificatesTests.cs [hunk 1]
```diff
@@ -19,7 +19,104 @@ namespace Halibut.Tests
...
+                serviceLoggers[serviceLoggers.Keys.First()].GetLogs().Should()
+                    .Contain(log => log.FormattedMessage
+                        .Contains("and attempted a message exchange, but it presented a client certificate with the thumbprint " +
+                                  "'76225C0717A16C1D0BA4A7FFA76519D286D8A248' which is not in the list of thumbprints that we trust"));
+            }
+        }
+
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testWebSocket: false, testPolling: false, testNetworkConditions: false, testAsyncAndSyncClients: true)]
+        public async Task SucceedWhenUsingSameConnection_EvenThoughClientCertificateHasHadTrustRemoved(ClientAndServiceTestCase clientAndServiceTestCase)
```
Comment (sburmanoctopus) [inline source/Halibut.Tests/BadCertificatesTests.cs:88]: This is the behaviour the code exhibits, so we wrote a test for it.

File: source/Halibut/HalibutRuntimeBuilder.cs [hunk 2]
```diff
@@ -74,6 +75,12 @@ internal HalibutRuntimeBuilder WithPollingReconnectRetryPolicy(Func<RetryPolicy>
             return this;
         }
 
+        public HalibutRuntimeBuilder WithOnUnauthorizedClientConnect(Func<string, string, UnauthorizedClientConnectResponse> onUnauthorizedClientConnect)
```
Comment (sburmanoctopus) [inline source/Halibut/HalibutRuntimeBuilder.cs:78]: This isn't needed, but it feels like we should start pushing OnUnauthorizedClientConnect to the builder (and eventually remove the public 'set' for this callback)

File: source/Halibut.Tests/BadCertificatesTests.cs [hunk 3]
```diff
@@ -28,7 +125,7 @@ public async Task FailWhenClientPresentsWrongCertificateToListeningService(Clien
             using (var clientAndBuilder = await clientAndServiceTestCase.CreateTestCaseBuilder()
                        .AsLatestClientAndLatestServiceBuilder()
                        .WithServiceTrustingTheWrongCertificate()
-                       .WithCountingService()
+                       .WithCountingService(countingService)
```
[the remark this answers is not in the document]
Comment (LukeButters) [reply source/Halibut.Tests/BadCertificatesTests.cs:161]: This should be fixed in a different way

[the remark this answers is not in the document]
Comment (LukeButters) [reply source/Halibut.Tests/BadCertificatesTests.cs:161]: Nice

**Review by LukeButters — APPROVED** [review]

**Comment by sburmanoctopus** [comment]
[sc-53157]

**Pull request merged**